src/spi2_pool.rs

```diff
@@ -0,0 +1,211 @@
+                SPI_DEVICE_NO_DUMMY
+            } else {
+                0_u32
+            },
+            ..Default::default()
+        }
+    }
+
+    fn rm_from_bus(&self) -> Result<(), EspError> {
+        for (id, handle) in &self.shared_handles {
+            let inner: spi_device_handle_t = *handle;
+            esp!(unsafe { spi_bus_remove_device(inner) })?;
+        }
+        Ok(())
+    }
+}
+
+impl<'d, T> Drop for SpiConfigPool<'d, T> {
+    fn drop(&mut self) {
+        self.rm_from_bus().unwrap();
+    }
+}
+
+pub struct SpiPoolDevice<'d, T> {
+    pool: T,
+    pub pin_driver: PinDriver<'d, AnyOutputPin, Output>,
+    config_id: u32,
+    _p: PhantomData<&'d ()>,
+}
+
+impl<'d, T> SpiPoolDevice<'d, T>
+where
+    T: Borrow<SpiConfigPool<'d, &'d SpiMasterDriver<'d>>> + 'd,
+{
+    pub fn new(
+        pool: T,
+        config_id: u32,
+        cs_pin: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
+    ) -> Result<Self, EspError> {
+        let cs: PeripheralRef<AnyOutputPin> = cs_pin.into_ref().map_into();
+        let mut pin_driver = PinDriver::output(cs)?;
+        pin_driver.set_high()?;
+        Ok(Self {
+            pool,
+            pin_driver,
+            config_id,
+            _p: PhantomData,
+        })
+    }
+
+    pub fn transaction<R, E>(
+        &mut self,
+        f: impl FnOnce(&mut SpiBusMasterDriver<'d>) -> Result<R, E>,
+    ) -> Result<R, E>
+    where
+        E: From<EspError>,
+        //X: Borrow<&SpiMasterDriver<'d>> + 'd,
+    {
+        // get config_handle for our device
+        let pool: &SpiConfigPool<&SpiMasterDriver> = self.pool.borrow();
+        // create error here when device has wrong id
+        let device_handle:&spi_device_handle_t = pool.shared_handles.get(&self.config_id).unwrap();
+
+        // lock the bus through the mutex in SpiMasterDriver
+        let master: &SpiMasterDriver = pool.master.borrow();
+        // should we check for poison here?
+        let driver_lock = master.handle.lock().unwrap();
+
+        //self.pin_driver.set_low()?;
+
+        //let handle = self.handle.clone().into_inner().unwrap();
+        let mut bus = SpiBusMasterDriver {
+            handle: *device_handle,
+            trans_len: SOC_SPI_MAXIMUM_BUFFER_SIZE as usize,
+            hardware_cs: false,
+            _p: PhantomData,
+        };
+        self.pin_driver.set_low()?;
+        let trans_result = f(&mut bus);
+
+        let finish_result = bus.finish();
+
+        // Flush whatever is pending.
+        // Note that this is done even when an error is returned from the transaction.
+        let flush_result = bus.flush();
+
+        self.pin_driver.set_high()?;
+
+        drop(driver_lock);
+        println!("after lock drop");
+        let result = trans_result?;
+        println!("after trans result");
+        flush_result?;
+        println!("flush and finish");
+        finish_result?;
+        println!("end Transaction");
+        Ok(result)
+    }
+
```

Comment:
With the change in my first comment, you'd be able to simplify this whole function to just this.
```suggestion
        let device = &pool.shared_handles.get(&self.config_id).unwrap();

        device.transaction(f);
```

Reply:
as stated above. this is much nicer but i need the difference from EspSpiDevice because this is software_cs. Software_cs need a dfferent SpiBusMasterDriver flag + needs tight setting over the pins. Have to think how to still  reduce some of the duplication efficiently.

Reply:
Understood